A-1: Exploration of data and description of content sets 

In [ ]:
import pandas as pd
import numpy as np


In [25]:
impressions_columns = ['COOKIE_ID','AD_ID','PLACEMENT_ID']
impressions_array = numpy.loadtxt('C:/Users/insan/PycharmProjects/Essence/data/impressions.txt')
impressions = pd.DataFrame(data = impressions_array, columns=impressions_columns)
print (impressions.head(n=10))
print ("Number of rows in IMPRESSIONS file is",len(impressions))

   COOKIE_ID  AD_ID  PLACEMENT_ID
0    12270.0  325.0        3719.0
1    15956.0  324.0        3718.0
2     3793.0  325.0        3714.0
3     9599.0  324.0        3721.0
4    16807.0  323.0        3720.0
5    12500.0  325.0        3719.0
6     1115.0  325.0        3714.0
7    16281.0  325.0        3722.0
8    16802.0  325.0        3714.0
9    16804.0  325.0        3727.0
Number of rows in IMPRESSIONS file is 40500


IMPRESSIONS.TXT identifies the ad shown to each user, and where that ad was placed, as tracked by cookies.

In [21]:
activity_columns = ['COOKIE_ID','ACTIVITY_TAG','ACTIVITY_TYPE']

activity_dtype = {'COOKIE_ID':float, 'ACTIVITY_TAG':object,'ACTIVITY_TYPE':object}
activity = pd.read_table('C:/Users/insan/PycharmProjects/Essence/data/activity.txt', header=None, names=activity_columns,dtype=activity_dtype)
print (activity.head(n=10))
print ("Number of rows in ACTIVITY file is",len(activity))


   COOKIE_ID ACTIVITY_TAG ACTIVITY_TYPE
0    11592.0     pur_conf             1
1     6289.0     pur_conf             1
2    10519.0     pur_conf             1
3    19695.0     pur_conf             2
4     4832.0     pur_conf             1
5    12950.0     pur_conf             1
6     9745.0     pur_conf             1
7    12089.0     pur_conf             2
8    14594.0     pur_conf             2
9     4043.0     pur_conf             1
Number of rows in ACTIVIT file is 3606


ACTIVITY.TXT contains data that indicates if the user completed a purchase, and whether that purchase was completed with or without an ad.  The dataset contains 3,606 rows.  

In [24]:
ad_lookup_columns = ['AD_ID','AD_SIZE','AD_FORMAT']
ad_lookup_dtype = {'AD_ID':float, 'AD_SIZE':object,'AD_FORMAT':object}
ad_lookup = pd.read_table('C:/Users/insan/PycharmProjects/Essence/data/ad_lookup.txt', header=None, names=ad_lookup_columns,dtype=ad_lookup_dtype)
#activity = pd.DataFrame(data = activity_array, columns=activity_columns])
ad_lookup.head(n=10)


<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>AD_ID</th>
      <th>AD_SIZE</th>
      <th>AD_FORMAT</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>323.0</td>
      <td>300x250</td>
      <td>Standard</td>
    </tr>
    <tr>
      <th>1</th>
      <td>324.0</td>
      <td>728x90</td>
      <td>Standard</td>
    </tr>
    <tr>
      <th>2</th>
      <td>325.0</td>
      <td>300x600</td>
      <td>Standard</td>
    </tr>
    <tr>
      <th>3</th>
      <td>326.0</td>
      <td>970x250</td>
      <td>Rich-Media</td>
    </tr>
    <tr>
      <th>4</th>
      <td>327.0</td>
      <td>500x600</td>
      <td>Rich-Media</td>
    </tr>
    <tr>
      <th>5</th>
      <td>328.0</td>
      <td>:15s</td>
      <td>Video</td>
    </tr>
  </tbody>
</table>
</div>

AD_LOOKUP contains the domain values which specify the various ad characteristics, such as size and format of ad.  IT contains 5 rows, and links to the IMPRESSSIONS  data through AD_ID


In [115]:
placement_lookup_columns = ['PLACEMENT_ID','CATEGORY_ID','CPI']
##activity_array = numpy.loadtxt('C:/Users/insan/PycharmProjects/Essence/data/activity.txt')
placement_lookup_dtype = {'PLACEMENT_ID':float, 'CATEGORY_ID':float,'CPI':float}
placement_lookup = pd.read_table('C:/Users/insan/PycharmProjects/Essence/data/placement_lookup.txt', header=None, names=placement_lookup_columns,dtype=placement_lookup_dtype)
#activity = pd.DataFrame(data = activity_array, columns=activity_columns])
print (placement_lookup.head(n=10))
print ("Length of PLACEMENT data is ",len(placement_lookup))


   PLACEMENT_ID  CATEGORY_ID   CPI
0        3712.0      10001.0  12.0
1        3713.0      10001.0  12.0
2        3714.0      10001.0  12.0
3        3715.0      10001.0  20.0
4        3716.0      10001.0  20.0
5        3717.0      10002.0   8.0
6        3718.0      10002.0   8.0
7        3719.0      10002.0   8.0
8        3720.0      10003.0  20.0
9        3721.0      10003.0  20.0
Length of PLACEMENT data is  16


PLACEMENT_LOOKUP contains details of the ad placement such as category and cost.  It contains 16 rows is tied to IMPRESSIONs via placement_id.


In [33]:
category_lookup_columns = ['CATEGORY_ID','CATEGORY','AVG_PURCHASE_PRICE']
category_lookup_dtype = {'CATEGORY_ID':float, 'CATEGORY':object,'AVG_PURCHASE_PRICE':float}
category_lookup = pd.read_table('C:/Users/insan/PycharmProjects/Essence/data/category_lookup.txt', header=None, names=category_lookup_columns,dtype=category_lookup_dtype)
category_lookup.head(n=10)

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CATEGORY_ID</th>
      <th>CATEGORY</th>
      <th>AVG_PURCHASE_PRICE</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>10001.0</td>
      <td>Travel</td>
      <td>200.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>10002.0</td>
      <td>Weather</td>
      <td>150.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>10003.0</td>
      <td>News</td>
      <td>250.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>10004.0</td>
      <td>Gaming</td>
      <td>50.0</td>
    </tr>
  </tbody>
</table>
</div>

CATEGORY_LOOKUP defines the  lookup values for website category data.  It contains 4 rows, and links to PLACEMENT_LOOKUP via placement_id.


In [ ]:
combined_placement = pd.merge(placement_lookup, category_lookup,on='CATEGORY_ID')
combined_impressions = pd.merge(impressions,activity, how='outer', on = 'COOKIE_ID')
combined_impressions=pd.merge(combined_impressions,ad_lookup,on="AD_ID")
combined_impressions = pd.merge(combined_impressions,combined_placement, on = 'PLACEMENT_ID')


print(combined_impressions.head(n=10))
print (len(combined_impressions))
#combined_impressions.to_csv("C:/Users/insan/PycharmProjects/Essence/data/combined_impressions.csv")


A-3: With the given data,  a number of potential measures of metrics could be used to assess activity at the cookie level; most basically, effectiveness ( as measured by conversion rate ) and efficiency (cost per purchase action) can be assessed by ad, placement, ad size, ad format, and website category with the given data.  Further analysis by revenue generated per action (accounting for ad cost) would also help to understand whether the ad purchase is financially sensible relative to other options.

In [120]:
#combined_impressions.query('ACTIVITY_TYPE'=='2').groupby('AD_ID').size().head(n=10)
ads_converted = combined_impressions[combined_impressions['ACTIVITY_TYPE']=='2'].groupby('AD_ID').size().to_frame('size_converted')
ads_cost = combined_impressions[combined_impressions['ACTIVITY_TYPE']=='2'].groupby('AD_ID')['CPI'].agg('sum').to_frame('CPI_TOTAL')
ads_cost['AD_ID']=ads_cost.index
ads_converted['AD_ID']=ads_converted.index
ads_total = combined_impressions.groupby('AD_ID')['AD_ID'].size().to_frame('total_impressions')
ads_total['AD_ID']=ads_total.index
ads_summary=pd.merge(ads_total,ads_converted,on='AD_ID')
ads_summary=pd.merge(ads_summary,ads_cost,on='AD_ID')
ads_summary['conversion_rate']=ads_summary['size_converted']/ads_summary['total_impressions']
ads_summary['efficiency']=ads_summary['CPI_TOTAL']/ads_summary['size_converted']


print (ads_summary)

Based on determining cconversion rate as  ratio of impressions with click-through activity to total number of impressions per ad, ad 326 demonstrates highest conversion rate amongst all ads, with conversion rate of 16%.   This effectiveness might be driven by the fact that this ad was only displayed on travel sites, and that the ad was tailored for consumers interested in travel content.

To determine efficiency, we assess dollar spent per purchase action (as generated by click-through on the ad).  With $8.85 spent per purchase, ad 324 is the most efficient; 73% of impressions were made on low-CPI category sites like travel and news.

<div>
<style>
    .dataframe thead tr:only-child th {
        text-align: right;
    }

    .dataframe thead th {
        text-align: left;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>size_converted</th>
      <th>AD_ID</th>
    </tr>
    <tr>
      <th>AD_ID</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>323.0</th>
      <td>11700</td>
      <td>323.0</td>
    </tr>
    <tr>
      <th>324.0</th>
      <td>11700</td>
      <td>324.0</td>
    </tr>
    <tr>
      <th>325.0</th>
      <td>11700</td>
      <td>325.0</td>
    </tr>
    <tr>
      <th>326.0</th>
      <td>1800</td>
      <td>326.0</td>
    </tr>
    <tr>
      <th>327.0</th>
      <td>3000</td>
      <td>327.0</td>
    </tr>
    <tr>
      <th>328.0</th>
      <td>600</td>
      <td>328.0</td>
    </tr>
  </tbody>
</table>
</div>

In [122]:
#combined_impressions.query('ACTIVITY_TYPE'=='2').groupby('AD_ID').size().head(n=10)
placement_converted = combined_impressions[combined_impressions['ACTIVITY_TYPE']=='2'].groupby('PLACEMENT_ID').size().to_frame('size_converted')
placement_cost = combined_impressions[combined_impressions['ACTIVITY_TYPE']=='2'].groupby('PLACEMENT_ID')['CPI'].agg('sum').to_frame('CPI_TOTAL')
placement_cost['PLACEMENT_ID']=placement_cost.index
placement_converted['PLACEMENT_ID']=placement_converted.index
placement_total = combined_impressions.groupby('PLACEMENT_ID')['PLACEMENT_ID'].size().to_frame('total_impressions')
placement_total['PLACEMENT_ID']=placement_total.index
placement_summary=pd.merge(placement_total,placement_converted,on='PLACEMENT_ID')
placement_summary=pd.merge(placement_summary,placement_cost,on='PLACEMENT_ID')
placement_summary['conversion_rate']=placement_summary['size_converted']/placement_summary['total_impressions']
placement_summary['efficiency']=placement_summary['CPI_TOTAL']/placement_summary['size_converted']


print (placement_summary)

    total_impressions  PLACEMENT_ID  size_converted  CPI_TOTAL  \
0                2400        3712.0             192     2304.0   
1                2400        3713.0              96     1152.0   
2                2400        3714.0             384     4608.0   
3                1800        3715.0             288     5760.0   
4                1800        3716.0             288     5760.0   
5                4500        3717.0             450     3600.0   
6                4500        3718.0             270     2160.0   
7                4500        3719.0             630     5040.0   
8                1800        3720.0             108     2160.0   
9                1800        3721.0              72     1440.0   
10               1800        3722.0             144     2880.0   
11               1200        3723.0              72     2016.0   
12                600        3724.0              36     1728.0   
13               3000        3725.0             180      720.0   
14        

Placements 3714,3715, and 3716 show highest effectiveness at 16%; high effeciveness may be driven by the fact that these placements were on travel sites, which are highly relevant to the content being sold (noting though that placements 3712 and 3713 were also on travel sites, but were not as effective).  3725, 3726, and 3727 show highest efficiency at $4 per ad-driven purchase; these placements were in gaming sites, which offer particularly low CPI.

B-3: Confounds: this analysis is driven by the assessment of the success of click-through conversion ads and sets aside those purchases made through view-through  
conversions, on the presumption that only purchases made through direct ad clicks are successful conversions.  This presumption discounts the potential value o
of ad views, which may not generate an immediate purchase, but may impact purchasing decisions at a future time.  

B-4: To improve the analysis, I would try to more fairly assess the impact of view-through conversions, which is likely a more significant driver of purchase activity than considered here.  In terms of additional data, information on the drivers of the CPI (such as placement website traffic data) would be useful to assess the "fairness" of the ad price.  I would also like additional details on the quality of the ad placement--where on webpage the ad is being placed, if scrolling is required to access the ad, if the ad is accessible on the main page vs sub-pages on the site etc.)

B-5: Based on this limited analysis, 